include/eosio/vm/argument_proxy.hpp

```diff
@@ -0,0 +1,120 @@
+#pragma once
+
+#include <eosio/vm/types.hpp>
+#include <eosio/vm/span.hpp>
+
+#include <memory>
+namespace eosio { namespace vm {
+
+   // Used for pointer arguments to intrinsics.
+   // T should be a type that points to external memory, such as a pointer or span.
+   // The type that T points to must be trivially copyable.
+   // argument_proxy copies unaligned memory into zero or more objects of the pointee type.
+   // If LegacyAlign is non-zero it will only copy the memory if it is not aligned to LegacyAlign.
+   // If the pointee is mutable, the memory will be written back by the destructor.
+   template <typename T, std::size_t LegacyAlign=0>
+   struct argument_proxy;
+
+   // specialization of argument_proxy for pointers.
+   // copies a single element.
+   template <typename T, std::size_t LegacyAlign>
+   struct argument_proxy<T*, LegacyAlign> {
+      static_assert(LegacyAlign % alignof(T) == 0, "Specified alignment must be at least alignment of T");
+      static_assert(std::is_trivially_copyable_v<T>, "argument_proxy requires a trivially copyable type");
+      inline constexpr argument_proxy(void* ptr) : original_ptr(ptr) {
+         if (!LegacyAlign || reinterpret_cast<std::uintptr_t>(original_ptr) % LegacyAlign != 0) {
+            copy.emplace();
+            memcpy( std::addressof(*copy), original_ptr, sizeof(T) );
+         }
+      }
+      inline constexpr argument_proxy(const argument_proxy&) = delete;
+      inline constexpr argument_proxy(argument_proxy&& other) : original_ptr(other.original_ptr), copy(other.copy) {
+         other.copy.reset();
+         other.original_ptr = nullptr;
+      }
+      inline ~argument_proxy() {
+         if constexpr (!std::is_const_v<T>)
+            if (copy)
+               memcpy( original_ptr, std::addressof(*copy), sizeof(T) );
+      }
+      constexpr operator T*() { return get(); }
+      constexpr operator const T*() const { return get(); }
+
+      constexpr T* get() { return const_cast<T*>(const_cast<const argument_proxy*>(this)->get()); }
+      constexpr const T* get() const {
+         if (copy)
+            return std::addressof(*copy);
+         else
+            return static_cast<const T*>(original_ptr);
+      }
+      constexpr T* operator->() { return get(); }
+      constexpr const T* operator->() const { return get(); }
+
+      static constexpr bool is_legacy() { return LegacyAlign != 0; }
+      using dependent_type = T;
```

Comment:
I think this should also add `using proxy_type = T*;` to make it easier to determine what kind of `argument_proxy` it is elsewhere.